### PR TITLE
add tags to flag structs

### PR DESCRIPTION
**Pull request by chippyash**
This adds the ability to set a flag name (tag) if there is an annotated struct. This is a partial step towards the ability to use real flag names, unrestricted by Go.  I cannot see where the 'getting' from the service of a flag value is done, as there are no tests to get intuition from.  However, I've amended tests as required and added new tests to show that Tag is a thing. I need help to get your service to use the tag values as the flag names.

### CODE REVIEW

**Comment by AsafRollout** [comment]
hi @chippyash 
thank you, for taking the time to improve our SDK
do you mean separating the name in the code (the Go field name), and the name in the Dashboard?
so on the SDK side everything will be with the (Go variable's) name, and when you see it in the Dashboard it will be the tag?
there are 2 direction for this:
1. configuration fetch from server, this will set for each flag the condition (experiment repository [file fetched from network [created with Dashboard names] with all flags data] to flag repository [the local variants being evaluated]), this is done here: [GetExperimentByFlag](https://github.com/rollout/rox-go/blob/914b422bf68f504fb12501a8fd6d3b992d585aeb/v5/core/entities/flag_setter.go#L25)
2. state sent to server, in order to create flag automatically (done here: [serializeFeatureFlags](https://github.com/rollout/rox-go/blob/914b422bf68f504fb12501a8fd6d3b992d585aeb/v4/core/network/state_sender.go#L132) )

i suppose that you will want to change both places to use the tag (Dashboard name) instead of the name
however, this will cause a weird behavior adding a tag:
a new flag will be created (with the tag name), and you will have to manually clone the configuration to the old flag name (the variant name)
you can use redundancy in 1 using either flag tag or name, but this duplication sounds risky
what did you intend to do about that?

**Comment by chippyash** [comment]
Hi @AsafRollout 

Yes, I think you have gotten the intent.

Unlike other language implementations, PHP for instance, the Go implementation forces you to use structs. They have their limitations in this use case as struct attribute names *must* start with a capital letter in order to be publicly visible, and cannot contain dots or dashes. 

Having set up our flags system originally for use with PHP, we were able to use flag names such as

- api.some-such-thing
- vm.some-other-thing

These are simply unreachable using the Go sdk. I would have to rename all my flags in the console and change thousands of lines of code in PHP. As PHP is on it's way out as we convert over to Go, that is untenable.

My suggestion is to use tags in the Go struct definition.  My PR demonstrates that it can be done. I guess the rest is up to you guys, because, as I say, there is no test code around the actual server implementation that I can rely on to 
a/ figure out how it is done
b/ be assured that what I do doesn't break anything.

In short, instead of using the flag `name`, use the flag `tag` for all interactions between client and server.  My PR shows that `tag` will default to name if it's not in an annotation.

Does that clarify things?

Kind regards and Merry Christmas.

I'll be back in the new year.

**Comment by AsafRollout** [comment]
everything is reachable, can use dynamicApi 😄 

think we can use a RoxOption to enable tags, or it will break for someone that already have a tag (when they will upgrade)

i'll look to see if we can easily complete this, thank you!

**Comment by AsafRollout** [comment]
@chippyash
any reason you need the Tag (as a flag property)?
using only the register.go code change - changing the Name to be the Tag (or fallback to the name itself) seems perfect

everything works by Name, and stays the same
the only problem is backwards compatibility, and i think we can say it's a breaking change, and if somebody upgrade they will have to copy the configuration from the field name (code name) to the tag (flagName)

**Comment by AsafRollout** [comment]
i mean something like this
https://github.com/rollout/rox-go/pull/54
can be enough

**Comment by AsafRollout** [comment]
@chippyash 
merged the PR to replace the name with the tag `flagName:"<flagName>"` (copied the changes in the registerer, didn't keep a tag on the flag, instead just changed the name)

thank you, and hope it's good enough for what you are trying to do

**Comment by chippyash** [comment]
Hi @AsafRollout 
I look forward to seeing this finally merged and giving it a try.
Thanks
